### ISSUE TITLE
Second screen does not scale correctly with multi monitor view in full screen mode

### ISSUE THREAD

**Issue by naatje80** [TurboVNC/turbovnc#203]
Hello,
There currently seems to be an issue with the multi monitor full screen mode. The display on the second monitor is only partially filled.

We are starting the TigerVNC server (version 1.10.0) on a Centos 7 linux server and connect to the pc via Windows 10. 

This only seems to be an issue for the TurboVNC client on Windows: I did not encounter the same issue with a manually compiled version of TurboVNC client on Linux. Because the Linux TurboVNC client is a java based client, I also tried the TurboVNC java client on Windows. But this does not make any difference: the same scaling issues. I also tried the TigerVNC client for Windows, but this version seems to be working correctly most of the time (I had to same issue once, but when a reconnected to the same session I could not reproduce it). 

I also tried the same on a Windows 10 system of one of my colleagues (she has a different multi-monitor setup with different resolutions), but the result is the same.

Thank you in advance for your help.

**Comment by dcommander** [comment]
Does the same issue occur with our server?

**Comment by naatje80** [comment]
@dcommander Thank you for your quick reply. I just tried with the turbovnc server: and exactly the same issue.
I've also attached a screenshot the issue. In the first screen of the full screenshot you can ignore the black border: this is generated by the screenshot application due to the mismatch of the resolutions. However, on the secondary screen is not fully utilizing all space: it seems to be limited to the height of the primary screen.
![2019-12-09_09-26-17](https://user-images.githubusercontent.com/26620355/70419895-bd1cf780-1a66-11ea-8e11-ebb5e340fcc4.jpg)

**Comment by naatje80** [comment]
@dcommander Were you able to find a cause of the issue? One update: It looks like it is only able to resize to the the max. height of the primary monitor. If I ensure that both monitors are using the same resolution, it is working correctly.

**Comment by dcommander** [comment]
Sorry for the delay.  I'll look at this this week.

**Comment by naatje80** [comment]
No problem, thank you for the update.

**Comment by dcommander** [comment]
Now that I look at this more closely, it might not be a bug but rather a limitation.  So let me explain the limitation, and you can tell me whether the explanation fits.

Due to limitations of the RFB protocol, the TurboVNC Viewer must use a single window on the client in order to display the remote desktop, even if the remote desktop contains multiple virtual screens.  The TurboVNC Viewer will transmit information about the client screen layout to the TurboVNC (or TigerVNC) Server, thus allowing the server to configure the remote desktop's virtual screens such that the screen boundaries align with those of the client.  But if the client-side screens are of different resolutions, then there isn't much that the viewer can do.  Since it is using a single window, that window must be confined vertically by the smallest client screen.

A more advanced approach would be to use a separate window on the client for each server-side virtual screen (see also: #141.)  However, that would significantly increase the complexity of TurboVNC.  In addition to the complexity of managing multiple client windows, which would involve a wholesale re-architecture of the TurboVNC Viewer, it would be necessary to implement new server-side logic to ensure that RFB updates do not span multiple virtual screens.  I posit that TurboVNC does a better job than most VNC viewers of handling full-screen multi-screen, but this is one fundamental limitation that probably can't be overcome with the existing architecture.

I'm not sure why the Linux TurboVNC Viewer works properly in this regard, since it is also using a single window to display all server-side virtual screens.

**Comment by naatje80** [comment]
Thank you for your clarification. Our current solution with RealVNC is exactly what you describe: we connect two clients to the two different server processes, each with their own resolution. But this currently only works for the older build in Xserver, so we are running in to all kind of issues with applications that require a newer version of X.

This is one of the reasons why we decided to move to an opensource solution. Although we prefer the separate viewer method, the full screen multi-monitor setup would also have been a good alternative. But not every user has a multi-monitor setup with the equal resolutions/monitors, so that would be an issue for us.

I understand your explanation of the limitation of the RFB protocol. But I do not understand why it works with the Linux client. I just did another test: I also tried the combination TurboVNC server and TurboVNC Linux client: It also worked correctly with that setup.

I also compared the output of xrand:
```
Screen 0: minimum 32 x 32, current 3280 x 1200, maximum 32767 x 32767
VNC-0 connected primary 1680x1050+0+0 0mm x 0mm
......
VNC-1 connected 1600x1200+1680+0 0mm x 0mm
```
So here it correctly detects the resolutions.

But within the Windows viewer it is indeed limited to the max. height:
```
Screen 0: minimum 32 x 32, current 3280 x 1050, maximum 32767 x 32767
VNC-0 connected primary 1680x1050+0+0 0mm x 0mm
.....
VNC-1 connected 1600x1050+1680+0 0mm x 0mm
```
Manually changing the resolution with xrandr did not work:

- Within the Linux client: It automatically resized back to the required resolutions.
- Within Windows: It seems to do nothing for the secondary monitor, however I could now scroll 

through the primary display (a black bar was added to the bottom to extend it to the 1200 height).
So here are also noticeable differences. The command that I tried was: `xrandr --output VNC-1 --mode 1600x1200`
Hopefully this gives you some new possible ideas of the cause of this.

Do you know if TigerVNC is also using the RFB protocol? Because for that client it is working correctly. But we depend also on some other features of TurboVNC, so that is why we prefer to use your client.

Thank you again for your help.

**Comment by dcommander** [comment]
TigerVNC uses the same exact protocol as us, so I'm not sure why that viewer works and ours doesn't.  I'll investigate that, as well as why the Linux viewer is working.

**Comment by dcommander** [comment]
Testing with two monitors, primary = 1344 x 840 and secondary = 1920 x 1200 (to the right of primary), I have confirmed that our Linux viewer does indeed work as expected.  It's odd, though.  With debugging enabled (`-loglevel 100`), I note that the TurboVNC Viewer code is trying to create a 3264 x 840 window, but the actual size of the window that is created is 3264 x 1200.  In other words, the full-screen multi-screen window is the union of all screens, not the intersection.  This may be happening "behind the scenes" because the Linux TurboVNC Helper is setting the `_NET_WM_FULLSCREEN_MONITORS` atom to provide the window manager a hint that the full-screen window should span multiple monitors.  It appears that the TigerVNC Viewer is, likewise, creating the full-screen multi-screen window as the union of all screens rather than the intersection.  I'll try doing that as well.

**Comment by dcommander** [comment]
So it turns out that sizing the full-screen viewer window to the union of all monitors (rather than the intersection) is the right thing to do if and only if:
1. The server has Xinerama capabilities (otherwise, the server will have just one virtual screen, regardless of the number of physical screens that the client has.)  That generally means that TurboVNC >= 2.2 or TigerVNC are required.
2. Automatic desktop sizing is enabled in the viewer (otherwise, the viewer has no way to inform the server of its desired screen layout, so we have to assume that the server may have just one virtual screen.)

The logic behind Item 2 is straightforward, but I'm working on how to detect Item 1, since that is necessary for backward compatibility with TurboVNC 2.1.x and earlier.

**Comment by naatje80** [comment]
Great, thank for your detailed feedback and all your effort. U understand you need to implement this carefully for backward compatibility.

**Comment by dcommander** [comment]
It should work now.  I tested various configurations, including:
1. full-screen multi-screen spanning with multiple virtual screens (Xinerama) and automatic desktop resizing
2. full-screen multi-screen spanning with one virtual screen and manual desktop resizing
3. full-screen multi-screen spanning with one virtual screen, the TurboVNC 2.1.x server (no Xinerama support), and automatic desktop resizing

with Linux, macOS, and Windows clients.

Notes on various platforms:

- **Windows and Windows/Java:**  no known remaining issues
- **Mac:** As always, full-screen multi-screen spanning only works if "Displays have separate spaces" is turned off in the Mission Control applet in System Preferences.  I uncovered an unrelated bug whereby, if the viewer is displaying a full-screen multi-screen window and you switch away from it (using ⌘-Tab or Ctrl-← or Ctrl-→) and then back to it, suddenly the full-screen window resizes to just one screen.  I spent hours trying to figure out a way around that but couldn't.  The workaround is to press Ctrl-Alt-Shift-Z to restore the full-screen window to its multi-screen layout.  I recognize that that's sub-optimal, but we're in kind of a support blind alley for full-screen mode on Mac.  The only way to create a full-screen multi-screen window on that platform is to disable "Displays have separate spaces" and use the deprecated com.apple.eawt classes in Java to toggle full-screen mode.  Neither is well-supported, unfortunately.
- **Linux:** Oddly, whereas things worked properly on this platform without the patch, introducing the patch made things work improperly.  Suddenly the desktop image on both screens was shifted down.  So I just disabled the new code on Linux.  In general, Linux implementations of Java tend to do some weird things with full-screen window management, which is why I have to twiddle the X11 window manager atoms in the TurboVNC Helper in order to force the full-screen window to span multiple screens.  The desktop image in Cases 2 and 3 above is still shifted down, but that has been the case for a while.  That appears to be a similar issue to the one that prompted me to twiddle the aforementioned X11 WM atoms.  Specifically, even though I am setting the multi-screen window geometry correctly, and Java is reporting back that the window geometry is set correctly, at the X11 level it isn't correct.

Please test and let me know if anything is still broken.

**Comment by dcommander** [comment]
@naatje80 Have you had a chance to test this?  This 2.2.4 release is on hold pending your test results.

**Comment by naatje80** [comment]
@dcommander I'm sorry I did not responded before. It certainly was not a lack of appreciation for your work. But I was confused by the fact that you already closed both tickets. Normally I'm used to the fact that I should not continue in a closed ticket. I also did not want to create a new ticket just to inform you the issue was resolved. Sorry for that. 

But the issue is indeed resolved, it is working great now. I once had a strange issue during the test with a full screen session on one monitor (it stretched below the Windows bar), but after a reboot this issue was resolved. So I assume it was related to Windows 10 itself.

Thank you for your quick response in solving this issue.
